devdoc/channel_2.md

```diff
@@ -0,0 +1,72 @@
+# channel_2 requirements:
+
+## Overview
+
+The current implementation of `channel` has certain pitfalls. `channel_2` aims to address these pitfalls by making the following changes:
+
+1. Introduce a `channel_2_reset` API: This API will allow the user to reset the `channel_2` to a clean state. It will cause all pending operations to be abandoned.
+
+2. Remove `ASYNC_OP`: The current implementation of `channel` provides the user with an `ASYNC_OP` when an operation (`channel_push` or `channel_pull`) is initiated. The user can call `async_op_cancel` to do a best-effort cancellation of the operation. There are 2 issues with this:
+    - Complicates object lifetimes: The lifetime of the operation must be linked to the `ASYNC_OP`and the `channel`, since each can be independently destroyed by the user.
+    - Inconsistent callback order: Consider when there are 3 `PUSH` operations in the `channel`: [`op1`, `op2`, `op3`] with data [`data1`, `data2`, `data3`] respectively.
+    
+    The receiver calls `channel_pull` 3 times and the sender cancels `op2` concurrently.
+    
+    The callback for `op1` is called with `data1` and `CHANNEL_CALLBACK_RESULT_OK`. 
+
+    The callback for `op2` is called with `NULL`  and `CHANNEL_CALLBACK_RESULT_CANCELLED`.
+
+    The callback for `op3` is called with `data3` and `CHANNEL_CALLBACK_RESULT_OK`.
+
+    The receiver will now contain [`data1`, `data3`].
+
```

Comment:
This paragraph is saying "channel has a problem that when one operation is cancelled, the subsequent ones are not.

This  implies that the user of channel wants to preserve the order of the operations. That implies completing POPs in the order that they were PUSHed.

This document is not addressing the topic of ordering for channel.

"channel.md" has "order" in it one time:
"`channel_pull` and `channel_push` can be called from different threads in any order. The callbacks are executed on threadpool threads."

So, we're not defining anywhere even for channel (v1) that the order is of concern.

Now, if we're saying that we do need order to be preserved, I think we should be much crisper which order is to be preserved (the order in which channel sees the PUSH operations for example).

I think this has nothing to do with ASYNC_OP. ASYNC_OP gives a framework for an asynchronous operation which is also cancellable.